askama_derive/src/generator/filter.rs

```diff
@@ -0,0 +1,921 @@
+    ctx: &Context<'_>,
+    name: &str,
+    node: Span<'_>,
+    input_args: &'b [WithSpan<'a, Expr<'a>>],
+    filter_args: &'static [&'static FilterArgument],
+    collected_args: &mut [&'b WithSpan<'a, Expr<'a>>],
+) -> Result<(), CompileError> {
+    // invariant: the parser ensures that named arguments come after positional arguments
+    let mut arg_idx = 0;
+    for arg in input_args {
+        let (idx, value) = if let Expr::NamedArgument(arg_name, ref value) = **arg {
+            let Some(idx) = filter_args
+                .iter()
+                .enumerate()
+                .find_map(|(idx, arg)| (arg.name == arg_name).then_some(idx))
+            else {
+                return Err(ctx.generate_error(
+                    match filter_args.len() {
+                        1 => fmt_left!("filter `{name}` does not have any arguments"),
+                        _ => fmt_right!(
+                            "filter `{name}` does not have an argument `{}`{}",
+                            arg_name.escape_debug(),
+                            ItsArgumentsAre(filter_args),
+                        ),
+                    },
+                    arg.span(),
+                ));
+            };
+            (idx, &**value)
+        } else {
+            let idx = arg_idx;
+            arg_idx += 1;
+            (idx, arg)
+        };
+
+        let Some(collected_arg) = collected_args.get_mut(idx) else {
+            return Err(ctx.generate_error(
+                format_args!(
+                    "filter `{name}` accepts at most {} argument{}{}",
+                    filter_args.len() - 1,
+                    if filter_args.len() != 2 { "s" } else { "" },
+                    ItsArgumentsAre(filter_args),
+                ),
+                arg.span(),
+            ));
+        };
+        if !is_argument_placeholder(replace(collected_arg, value)) {
+            return Err(ctx.generate_error(
+                format_args!(
+                    "argument `{}` was already set",
+                    filter_args[idx].name.escape_debug()
+                ),
+                arg.span(),
+            ));
+        }
+    }
+
+    for (&arg, collected) in filter_args.iter().zip(collected_args) {
+        if !is_argument_placeholder(collected) {
+            continue;
+        } else if let Some(default) = arg.default_value {
+            *collected = default;
+        } else {
+            return Err(ctx.generate_error(
+                format_args!(
+                    "`{}` argument is missing when calling `{name}` filter{}",
+                    arg.name.escape_debug(),
+                    ItsArgumentsAre(filter_args),
+                ),
+                node,
+            ));
+        }
+    }
+
+    Ok(())
+}
+
+struct ItsArgumentsAre(&'static [&'static FilterArgument]);
+
+impl fmt::Display for ItsArgumentsAre {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("; it's arguments are: (")?;
```

Comment:
Missed this one too. ^^'

```suggestion
        f.write_str("; its arguments are: (")?;
```